x-pack/beatless/cmd/cli_handler.go

```diff
@@ -0,0 +1,131 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package cmd
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strings"
+
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/x-pack/beatless/core"
+	"github.com/elastic/beats/x-pack/beatless/provider"
+)
+
+// Errors generated by the cliHandler.
+var (
+	errNoFunctionGiven = errors.New("no function given")
+)
+
+// cliHandler takes a provider.CLIManager and acts a bridge between user enterred content from the CLI
+// and the type managing the function on the provider. It allow to specify multiple functions at
+// the command line but will do a single invocation on the CLIManager and will do general validation
+// and normalization of the values. It also communicate the status of the operations to the user.
+//
+// NOTES: Each execution call of the CLIManager are independant, this mean that a fail call will not
+// stop other calls to succeed.
+//
+// TODO(ph) functions could be merged into a single call , but I thought it was premature to do
+// it.
+type cliHandler struct {
+	cli       provider.CLIManager
+	log       *logp.Logger
+	errOutput io.Writer
+	output    io.Writer
+}
+
+func newCLIHandler(cli provider.CLIManager, errOutput io.Writer, output io.Writer) *cliHandler {
+	return &cliHandler{cli: cli, errOutput: errOutput, output: output, log: logp.NewLogger("cli-handler")}
+}
+
+func (c *cliHandler) Deploy(names []string) error {
+	c.log.Debugf("starting deploy for: %s", strings.Join(names, ", "))
+	defer c.log.Debug("deploy execution ended")
+
+	if len(names) == 0 {
+		return errNoFunctionGiven
+	}
+
+	errCount := 0
+	for _, name := range names {
+		if err := c.cli.Deploy(name); err != nil {
+			fmt.Fprintf(c.errOutput, "function: %s, could not deploy, error: %s\n", name, err)
+			errCount++
+			continue
+		}
+		fmt.Fprintf(c.output, "function: %s, deploy successful\n", name)
+	}
+
+	if errCount > 0 {
+		return fmt.Errorf("fail to deploy %d function(s)", errCount)
+	}
+	return nil
+}
```

Comment:
> Have you considered to run the operations in parallel?

I did consider it but I've decided against it to limit any concurrency issue I could get on the provider, but we could make the provider CLI implement a parallel or a serial interface and we could let the cli handler apply the right execution model. I won't do it for v1.

Reply:
I've used `fmt.Fprintf()` instead of a logger because theses are user visible error message like any command line. So there are not really log statements.